File: Source/Android/luascriptcore/src/main/java/cn/vimfung/luascriptcore/LuaValue.java
package cn.vimfung.luascriptcore;

import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Lua值对象
 * Created by vimfung on 16/8/23.
 */
public class LuaValue extends LuaBaseObject
{
    private Object _valueContainer;
    private LuaValueType _type;
    private LuaContext _context;    //从JNI层回来的对象会带上这个字段
    private String _tableId;           //Map和Array类型下的Table标识

    /**
     * 初始化一个空值的LuaValue对象
     */
    public LuaValue ()
    {
        super();
        setNilValue();
    }

    /**
     * 初始化一个控制的LuaValue对象
     * @param nativeId  本地对象标识
     */
    protected LuaValue(int nativeId)
    {
        super(nativeId);
        setNilValue();
    }

    /**
     * 设置空值
     */
    private void setNilValue ()
    {
        _type = LuaValueType.Nil;
        _valueContainer = null;
    }

    /**
     * 初始化一个整型值的LuaValue对象
     * @param value 整型值
     */
    public LuaValue(Integer value)
    {
        super();
        setLongValue(Long.valueOf(value));
    }

    /**
     * 初始化一个整型值的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 整型值
     */
    protected LuaValue(int nativeId, Integer value)
    {
        super(nativeId);
        setLongValue(Long.valueOf(value));
    }

    /**
     * 初始化一个长整型的LuaValue对象
     * @param value 长整型值
     */
    public LuaValue(Long value)
    {
        super();
        setLongValue(value);
    }

    /**
     * 设置长整型值
     * @param value 长整型值
     */
    private void setLongValue(Long value)
    {
        _type = LuaValueType.Integer;
        _valueContainer = value;
    }

    /**
     * 初始化一个浮点值的LuaValue对象
     * @param value 浮点数值
     */
    public LuaValue(Double value)
    {
        super();
        setDoubValue(value);

    }

    /**
     * 初始化一个浮点值的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 浮点数值
     */
    protected LuaValue(int nativeId, Double value)
    {
        super(nativeId);
        setDoubValue(value);
    }

    /**
     * 设置双精度浮点型数值
     * @param value 浮点数值
     */
    private void setDoubValue(Double value)
    {
        _type = LuaValueType.Number;
        _valueContainer = value;
    }

    /**
     * 初始化一个布尔值的LuaValue对象
     * @param value 布尔值
     */
    public LuaValue(Boolean value)
    {
        super();
        setBoolValue(value);
    }

    /**
     * 初始化一个布尔值的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 布尔值
     */
    protected LuaValue(int nativeId, Boolean value)
    {
        super(nativeId);
        setBoolValue(value);
    }

    /**
     * 设置布尔值
     * @param value 布尔值
     */
    private void setBoolValue(Boolean value)
    {
        _type = LuaValueType.Boolean;
        _valueContainer = value;
    }

    /**
     * 初始化一个字符串的LuaValue对象
     * @param value 字符串
     */
    public LuaValue(String value)
    {
        super();
        setStringValue(value);
    }

    /**
     * 初始化一个字符串的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 字符串
     */
    protected LuaValue(int nativeId, String value)
    {
        super(nativeId);
        setStringValue(value);
    }

    /**
     * 设置字符串
     * @param value 字符串
     */
    private void setStringValue(String value)
    {
        _type = LuaValueType.String;
        _valueContainer = value;
    }

    /**
     * 初始化一个字节数组的LuaValue对象
     * @param value 字节数组
     */
    public LuaValue (byte[] value)
    {
        super();
        setByteArrayValue(value);
    }

    /**
     * 初始化一个字节数组的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 字节数组
     */
    protected LuaValue(int nativeId, byte[] value)
    {
        super(nativeId);
        setByteArrayValue(value);
    }

    /**
     * 设置二进制数组
     * @param value 二进制数组值
     */
    private void setByteArrayValue(byte[] value)
    {
        _type = LuaValueType.Data;
        _valueContainer = value;
    }

    /**
     * 初始化一个字节数组的LuaValue对象
     * @param value 字节数组
     */
    public LuaValue (Byte[] value)
    {
        super();
        setByteArrayValue(value);
    }

    /**
     * 设置二进制数组
     * @param value 二进制数组
     */
    private void setByteArrayValue(Byte[] value)
    {
        _type = LuaValueType.Data;
        _valueContainer = value;
    }

    /**
     * 初始化一个数组的LuaValue对象
     * @param value 数组
     */
    public LuaValue (List<?> value)
    {
        super();
        setArrayListValue(value);
    }

    /**
     * 初始化一个数组的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 数组
     */
    protected LuaValue (int nativeId, List<?> value, String tableId)
    {
        super(nativeId);
        setArrayListValue(value);
        _tableId = tableId;
    }

    /**
     * 设置数组列表
     * @param value 数组列表值
     */
    private void setArrayListValue(List<?> value)
    {
        _type = LuaValueType.Array;
        _valueContainer = value;
    }

    /**
     * 初始化一个哈希表的LuaValue对象
     * @param value 哈希表
     */
    public LuaValue (Map<?, ?> value)
    {
        super();
        setHasMapValue(value);
    }

    /**
     * 初始化一个哈希表的LuaValue对象
     * @param nativeId 本地对象标识
     * @param value 哈希表
     */
    protected LuaValue (int nativeId, Map<?, ?> value, String tableId)
    {
        super(nativeId);
        setHasMapValue(value);
        _tableId = tableId;
    }

    /**
     * 设置HashMap
     * @param value HashMap值
     */
    private void setHasMapValue (Map<?, ?> value)
    {
        _type = LuaValueType.Map;
        _valueContainer = value;
    }

    /**
     * 初始化指针对象的LuaValue对象
     * @param value 指针对象
     */
    public LuaValue (LuaPointer value)
    {
        super();
        setPointerValue(value);
    }

    /**
     * 初始化指针对象的LuaValue对象
     * @param nativeId  本地标识对象
     * @param value 指针对象
     */
    protected LuaValue (int nativeId, LuaPointer value)
    {
        super(nativeId);
        setPointerValue(value);
    }

    /**
     * 设置指针对象
     * @param value 指针对象
     */
    private void setPointerValue (LuaPointer value)
    {
        _type = LuaValueType.Pointer;
        _valueContainer = value;
    }

    /**
     * 初始化值对象
     * @param value 方法对象
     */
    public LuaValue (LuaFunction value)
    {
        super();
        setFunctionValue(value);
    }

    /**
     * 初始化值对象
     * @param nativeId  本地标识
     * @param value     方法对象
     */
    protected  LuaValue (int nativeId, LuaFunction value)
    {
        super(nativeId);
        setFunctionValue(value);
    }

    /**
     * 设置方法对象
     * @param value 方法对象
     */
    private void setFunctionValue (LuaFunction value)
    {
        _type = LuaValueType.Function;
        _valueContainer = value;
    }

    /**
     * 设置元组
     * @param value 元组对象
     */
    public LuaValue (LuaTuple value)
    {
        super();
        setTupleValue(value);
    }

    /**
     * 初始化值对象
     * @param nativeId  本地标识
     * @param value     元组对象
     */
    protected  LuaValue (int nativeId, LuaTuple value)
    {
        super(nativeId);
        setTupleValue(value);
    }

    /**
     * 设置元组值
     * @param value 元组对象
     */
    private void setTupleValue (LuaTuple value)
    {
        _type = LuaValueType.Tuple;
        _valueContainer = value;
    }

    /**
     * 初始化值对象
     * @param value 对象
     */
    public LuaValue (Object value)
    {
        super();
        setObjectValue(value);
    }

    /**
     * 初始化值对象
     * @param nativeId  本地标识
     * @param value 对象
     */
    protected LuaValue (int nativeId, Object value)
    {
        super(nativeId);
        setObjectValue(value);
    }

    /**
     * 设置对象值
     * @param value 对象
     */
    private void setObjectValue (Object value)
    {
        if (value != null)
        {
            if (value instanceof Character)
            {
                setLongValue(Long.valueOf((Character)value));
            }
            else if (value instanceof Byte)
            {
                setLongValue(Long.valueOf((Byte)value));
            }
            else if (value instanceof Short)
            {
               setLongValue(Long.valueOf((Short)value));
            }
            else if (value instanceof Integer)
            {
                setLongValue(Long.valueOf((Integer)value));
            }
            else if (value instanceof Long)
            {
                setLongValue((Long) value);
            }
            else  if (value instanceof Float)
            {
                setDoubValue(((Float) value).doubleValue());
            }
            else if (value instanceof Double)
            {
                setDoubValue((Double)value);
            }
            else if (value instanceof Boolean)
            {
                setBoolValue((Boolean)value);
            }
            else if (value instanceof String)
            {
                setStringValue(value.toString());
            }
            else if (value instanceof byte[])
            {
                setByteArrayValue((byte[]) value);
            }
            else if (value instanceof List<?>)
            {
                setArrayListValue((List<?>)value);
            }
            else if (value instanceof Map<?, ?>)
            {
                setHasMapValue((Map<?, ?>)value);
            }
            else if (value instanceof LuaPointer)
            {
                setPointerValue((LuaPointer) value);
            }
            else if (value instanceof LuaFunction)
            {
                setFunctionValue((LuaFunction) value);
            }
            else if (value instanceof LuaTuple)
            {
                setTupleValue((LuaTuple) value);
            }
            else
            {
                _type = LuaValueType.Object;
                _valueContainer = value;
            }
        }
        else
        {
            _type = LuaValueType.Nil;
            _valueContainer = null;
        }
    }

    /**
     * 获取值类型
     * @return  类型
     */
    public LuaValueType valueType ()
    {
        return _type;
    }

    /**
     * 转换为整数
     * @return 整数
     */
    public long toInteger()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof Number)
            {
                return ((Number) _valueContainer).longValue();
            }
            else if (_valueContainer instanceof String)
            {
                return Integer.parseInt((String) _valueContainer);
            }
        }

        return 0;
    }

    /**
     * 转换为浮点数
     * @return 浮点数
     */
    public double toDouble()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof Number)
            {
                return ((Number) _valueContainer).doubleValue();
            }
            else if (_valueContainer instanceof String)
            {
                return Double.parseDouble((String) _valueContainer);
            }
        }

        return 0.0;
    }

    /**
     * 转换为布尔值
     * @return 布尔值
     */
    public boolean toBoolean()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof Boolean)
            {
                return (Boolean) _valueContainer;
            }
            else if (_valueContainer instanceof String)
            {
                return Boolean.parseBoolean((String) _valueContainer);
            }
        }

        return false;
    }

    /**
     * 转换为字符串
     * @return  字符串
     */
    public String toString()
    {
        if (_valueContainer != null)
        {
            return _valueContainer.toString();
        }

        return "";
    }

    /**
     * 转换为二进制数组
     * @return  二进制数组
     */
    public byte[] toByteArray()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof byte[])
            {
                return (byte[]) _valueContainer;
            }
            else if (_valueContainer instanceof Byte[])
            {
                final Byte[] srcBytes = (Byte[])_valueContainer;
                byte[] bytes = new byte[srcBytes.length];
                for (int i = 0; i < srcBytes.length; i++)
                {
                    bytes[i]= srcBytes[i];
                }

                return bytes;
            }
            else if (_valueContainer instanceof String)
            {
                return ((String) _valueContainer).getBytes();
            }
        }

        return null;
    }

    /**
     * 转换为数组列表
     * @return 数组列表
     */
    public ArrayList toArrayList()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof ArrayList)
            {
                return (ArrayList) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为列表
     * @return 列表对象
     */
    public List<?> toList()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof List<?>)
            {
                return (List<?>) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为哈希表
     * @return  哈希表
     */
    public HashMap toHashMap()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof HashMap)
            {
                return (HashMap) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为字典
     * @return 字典对象
     */
    public Map<?, ?> toMap()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof Map<?, ?>)
            {
                return (Map<?, ?>) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为指针对象
     * @return  指针对象
     */
    public LuaPointer toPointer()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof LuaPointer)
            {
                return (LuaPointer) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为方法对象
     * @return 方法对象
     */
    public LuaFunction toFunction()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof LuaFunction)
            {
                return (LuaFunction) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为元组对象
     * @return  方法对象
     */
    public LuaTuple toTuple()
    {
        if (_valueContainer != null)
        {
            if (_valueContainer instanceof LuaTuple)
            {
                return (LuaTuple) _valueContainer;
            }
        }

        return null;
    }

    /**
     * 转换为对象
     * @return  对象
     */
    public Object toObject()
    {
        return _valueContainer;
    }

    /**
     * 设置对象对象
     * @param keyPath  键名路径，如：key1.key2.key3
     * @param object 对象
     */
    public void setObject(String keyPath, Object object)
    {
        if (valueType() == LuaValueType.Map)
        {
            _valueContainer = LuaNativeUtil.luaValueSetObject(_context, this, keyPath, new LuaValue(object));
        }
    }

    /**
     * 获取table标识
     * @return table标识
     */
    protected String getTableId()
    {
        return _tableId;
    }
}

File: Source/Unity3D/UnityProject/Assets/Plugins/LuaScriptCore/LuaValue.cs
using System.Collections;
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;
using UnityEngine;

namespace cn.vimfung.luascriptcore
{
	public enum LuaValueType
	{
		Nil = 0,
		Number = 1,
		Boolean = 2,
		String = 3,
		Array = 4,
		Map = 5,
		Ptr = 6,
		Object = 7,
		Integer = 8,
		Data = 9,
		Function = 10,
		Tuple = 11
	};

	public class LuaValue : LuaBaseObject
	{
		private LuaValueType _type;
		private object _value;
		private string _tableId;
		private LuaContext _context;

		/// <summary>
		/// 初始化一个Nil值
		/// </summary>
		public LuaValue()
		{
			_value = null;
			_type = LuaValueType.Nil;
		}

		/// <summary>
		/// 初始化一个整型值
		/// </summary>
		/// <param name="value">整型值</param>
		public LuaValue(int value)
		{
			_value = value;
			_type = LuaValueType.Integer;
		}

		/// <summary>
		/// 初始化一个双精度浮点数
		/// </summary>
		/// <param name="value">浮点数</param>
		public LuaValue(double value)
		{
			_value = value;
			_type = LuaValueType.Number;
		}

		/// <summary>
		/// 初始化一个布尔值
		/// </summary>
		/// <param name="value">布尔值</param>
		public LuaValue(bool value)
		{
			_value = value;
			_type = LuaValueType.Boolean;
		}

		/// <summary>
		/// 初始化一个字符串值
		/// </summary>
		/// <param name="value">字符串.</param>
		public LuaValue(string value)
		{
			_value = value;
			_type = LuaValueType.String;
		}

		/// <summary>
		/// 初始化一个数组值
		/// </summary>
		/// <param name="value">数组</param>
		public LuaValue(List<LuaValue> value)
		{
			_value = value;
			_type = LuaValueType.Array;
		}

		/// <summary>
		/// 初始化一个字典值
		/// </summary>
		/// <param name="value">字典.</param>
		public LuaValue(Dictionary<string, LuaValue> value)
		{
			_value = value;
			_type = LuaValueType.Map;
		}

		/// <summary>
		/// 初始化一个二进制数组值
		/// </summary>
		/// <param name="value">二进制数组.</param>
		public LuaValue(byte[] value)
		{
			_value = value;
			_type = LuaValueType.Data;
		}

		/// <summary>
		/// 初始化一个方法值
		/// </summary>
		/// <param name="value">Lua方法.</param>
		public LuaValue (LuaFunction value)
		{
			_value = value;
			_type = LuaValueType.Function;
		}

		/// <summary>
		/// 初始化一个指针值
		/// </summary>
		/// <param name="value">Lua指针.</param>
		public LuaValue (LuaPointer value)
		{
			_value = value;
			_type = LuaValueType.Ptr;
		}

		/// <summary>
		/// 初始化一个元组值
		/// </summary>
		/// <param name="value">元组.</param>
		public LuaValue (LuaTuple value)
		{
			_value = value;
			_type = LuaValueType.Tuple;
		}

		/// <summary>
		/// 初始化一个对象值
		/// </summary>
		/// <param name="value">对象</param>
		public LuaValue (object value)
		{
			if (value != null)
			{
				
				if (value is int
				    || value is uint
				    || value is Int16
				    || value is UInt16
				    || value is Int64
				    || value is UInt64)
				{
					_value = value;
					_type = LuaValueType.Integer;
				}
				else if (value is double
				         || value is float)
				{
					_value = value;
					_type = LuaValueType.Number;
				}
				else if (value is bool)
				{
					_value = value;
					_type = LuaValueType.Boolean;
				}
				else if (value is byte[])
				{
					_value = value;
					_type = LuaValueType.Data;
				}
				else if (value is string)
				{
					_value = value;
					_type = LuaValueType.String;
				}
				else if (value is Array)
				{
					List<LuaValue> arr = new List<LuaValue> ();
					_value = arr;
					_type = LuaValueType.Array;

					//转换数据
					foreach (object item in (value as Array))
					{
						LuaValue itemValue = new LuaValue (item);
						arr.Add (itemValue);
					}
				}
				else if (value is IDictionary)
				{
					Dictionary<string, LuaValue> dict = new Dictionary<string, LuaValue> ();
					_value = dict;
					_type = LuaValueType.Map;

					foreach (DictionaryEntry de in (value as IDictionary))
					{
						LuaValue itemValue = new LuaValue (de.Value);
						dict.Add (Convert.ToString (de.Key), itemValue);
					}
				}
				else if (value is LuaFunction)
				{
					_value = value;
					_type = LuaValueType.Function;
				}
				else if (value is LuaPointer)
				{
					_value = value;
					_type = LuaValueType.Ptr;
				}
				else if (value is LuaTuple)
				{
					_value = value;
					_type = LuaValueType.Tuple;
				}
				else
				{
					if (value is LuaObjectDescriptor)
					{
						_value = value;
					}
					else if (value is ILuaObject)
					{
						_value = (value as ILuaObject).getDescriptor ();
					}
					else
					{
						_value = new LuaObjectDescriptor (value);
					}
					_type = LuaValueType.Object;
				}
			}
			else
			{
				_value = null;
				_type = LuaValueType.Nil;
			}
		}

		/// <summary>
		/// 读取一个数组列表
		/// </summary>
		/// <returns>数组列表</returns>
		/// <param name="decoder">对象解码器</param>
		private List<LuaValue> readArrayList(LuaObjectDecoder decoder)
		{
			int size = decoder.readInt32 ();
			if (size > 0) 
			{
				List<LuaValue> list = new List<LuaValue> ();
				for (int i = 0; i < size; i++)
				{
					LuaValue item = decoder.readObject () as LuaValue;
					if (item != null) 
					{
						list.Add (item);
					}
				}

				return list;
			}

			return null;
		}

		/// <summary>
		/// 读取一个字典
		/// </summary>
		/// <returns>字典对象</returns>
		/// <param name="decoder">对象解码器</param>
		private Dictionary<string, LuaValue> readHashtable(LuaObjectDecoder decoder)
		{
			int size = decoder.readInt32 ();
			if (size > 0) 
			{
				Dictionary<string, LuaValue> table = new Dictionary<string, LuaValue> ();
				for (int i = 0; i < size; i++) 
				{
					string key = decoder.readString ();
					LuaValue item = decoder.readObject () as LuaValue;
					if (key != null && item != null) 
					{
						table.Add (key, item);
					}
				}

				return table;
			}

			return null;
		}

		/// <summary>
		/// 序列化对象
		/// </summary>
		/// <param name="encoder">对象编码器.</param>
		public override void serialization (LuaObjectEncoder encoder)
		{
			base.serialization (encoder);

			if (_context != null)
			{
				encoder.writeInt32 (_context.objectId);
			}
			else
			{
				encoder.writeInt32 (0);
			}

			encoder.writeString (_tableId);
			encoder.writeInt16 ((Int16)type);

			switch (type)
			{
			case LuaValueType.Number:
				{
					encoder.writeDouble (toNumber ());
					break;
				}
			case LuaValueType.Integer:
				{
					encoder.writeInt32 (toInteger ());
					break;
				}
			case LuaValueType.String:
				{
					encoder.writeString (toString ());
					break;
				}
			case LuaValueType.Data:
				{
					encoder.writeBytes (toData ());
					break;
				}
			case LuaValueType.Array:
				{
					List<LuaValue> list = toArray ();

					encoder.writeInt32 (list.Count);
					foreach (LuaValue value in list) 
					{
						encoder.writeObject (value);
					}
					break;
				}
			case LuaValueType.Map:
				{
					Dictionary<string, LuaValue> map = toMap();
					encoder.writeInt32 (map.Count);

					foreach (KeyValuePair<string, LuaValue> kv in map) 
					{
						encoder.writeString (kv.Key);
						encoder.writeObject (kv.Value);
					}
					break;
				}
			case LuaValueType.Object:
				{
					encoder.writeObject (_value);
					break;
				}
			case LuaValueType.Boolean:
				{
					encoder.writeByte (Convert.ToByte(toBoolean ()));
					break;
				}
			case LuaValueType.Function:
				{
					encoder.writeObject (toFunction());
					break;
				}
			case LuaValueType.Ptr:
				{
					encoder.writeObject (toPointer());
					break;
				}
			case LuaValueType.Tuple:
				{
					encoder.writeObject (toTuple());
					break;
				}
			default:
				break;
			}
		}

		/// <summary>
		/// 获取值类型
		/// </summary>
		/// <value>类型</value>
		public LuaValueType type
		{
			get 
			{
				return _type;
			}
		}	

		/// <summary>
		/// 初始化LuaValue对象
		/// </summary>
		/// <param name="decoder">对象解码器</param>
		public LuaValue (LuaObjectDecoder decoder)
			: base(decoder)
		{

			int contextId = decoder.readInt32 ();
			_context = LuaContext.getContext (contextId);
			_tableId = decoder.readString ();
			_type = (LuaValueType)decoder.readInt16 ();
			_value = null;

			switch (_type) 
			{
			case LuaValueType.Integer:
				_value = decoder.readInt32 ();
				break;
			case LuaValueType.Boolean:
				_value = decoder.readByte ();
				break;
			case LuaValueType.Number:
				_value = decoder.readDouble ();
				break;
			case LuaValueType.Data:
				_value = decoder.readBytes ();
				break;
			case LuaValueType.String:
				_value = decoder.readString ();
				break;
			case LuaValueType.Array:
				_value = readArrayList (decoder);
				break;
			case LuaValueType.Map:
				_value = readHashtable (decoder);
				break;
			case LuaValueType.Function:
				_value = decoder.readObject () as LuaFunction;
				break;
			case LuaValueType.Ptr:
				_value = decoder.readObject () as LuaPointer;
				break;
			case LuaValueType.Tuple:
				_value = decoder.readObject () as LuaTuple;
				break;
			case LuaValueType.Object:
				_value = decoder.readObject ();
				break;
			}
		}

		/// <summary>
		/// 转换为整型
		/// </summary>
		/// <returns>整型值</returns>
		public int toInteger()
		{
			try
			{
				return Convert.ToInt32 (_value);
			}
			catch(FormatException ex)
			{
				Debug.LogFormat ("{0}", ex.Message);
				return 0;
			}
		}

		/// <summary>
		/// 转换为布尔值
		/// </summary>
		/// <returns>布尔值</returns>
		public bool toBoolean()
		{
			try
			{
				return Convert.ToBoolean (_value);
			}
			catch (FormatException ex)
			{
				Debug.LogFormat ("{0}", ex.Message);
				return false;
			}
		}

		/// <summary>
		/// 转换为双精度浮点型
		/// </summary>
		/// <returns>双精度浮点值</returns>
		public double toNumber()
		{
			try
			{
				return Convert.ToDouble (_value);
			}
			catch (FormatException ex)
			{
				Debug.LogFormat ("{0}", ex.Message);
				return 0.0;
			}

		}

		/// <summary>
		/// 转换为二进制数据流
		/// </summary>
		/// <returns>二进制数据流</returns>
		public byte[] toData()
		{
			try
			{
				switch (_type) 
				{
				case LuaValueType.Integer:
					return BitConverter.GetBytes (Convert.ToInt32 (_value));
				case LuaValueType.Number:
					return BitConverter.GetBytes (Convert.ToDouble(_value));
				case LuaValueType.String:
					return System.Text.Encoding.UTF8.GetBytes (Convert.ToString(_value));
				case LuaValueType.Data:
					return (byte[])_value;
				default:
					return null;
				}
			}
			catch (FormatException ex)
			{
				Debug.LogFormat ("{0}", ex.Message);
				return null;
			}
		}

		/// <summary>
		/// 转换为UTF8编码字符串
		/// </summary>
		/// <returns>字符串</returns>
		public string toString()
		{
			try
			{
				return Convert.ToString (_value);
			}
			catch (FormatException ex)
			{
				Debug.LogFormat ("{0}", ex.Message);
				return null;
			}
		}

		/// <summary>
		/// 转换为数组
		/// </summary>
		/// <returns>数组对象</returns>
		public List<LuaValue> toArray()
		{
			return _value as List<LuaValue>;
		}

		/// <summary>
		/// 转换为字典
		/// </summary>
		/// <returns>字典对象</returns>
		public Dictionary<string, LuaValue> toMap()
		{
			return _value as Dictionary<string, LuaValue>;
		}

		/// <summary>
		/// 转换为Lua方法
		/// </summary>
		/// <returns>Lua方法.</returns>
		public LuaFunction toFunction()
		{
			return _value as LuaFunction;
		}

		/// <summary>
		/// 转换为Lua指针
		/// </summary>
		/// <returns>Lua指针.</returns>
		public LuaPointer toPointer()
		{
			return _value as LuaPointer;
		}

		/// <summary>
		/// 转换为元组
		/// </summary>
		/// <returns>元组.</returns>
		public LuaTuple toTuple()
		{
			return _value as LuaTuple;
		}

		/// <summary>
		/// 转换为对象
		/// </summary>
		/// <returns>对象.</returns>
		public object toObject()
		{
			//解包数组和字典中的元素
			if (this.type == LuaValueType.Map)
			{
				Dictionary<string, LuaValue> map = toMap ();
				if (map != null)
				{
					Hashtable dict = new Hashtable ();
					foreach (KeyValuePair<string, LuaValue> kv in map)
					{
						dict.Add (kv.Key, kv.Value.toObject ());
					}
					return dict;
				}
			}
			else if (this.type == LuaValueType.Array)
			{
				List<LuaValue> arr = toArray();
				if (arr != null)
				{
					ArrayList objArr = new ArrayList ();
					foreach (LuaValue item in arr) 
					{
						objArr.Add (item.toObject ());
					}
					return objArr.ToArray ();
				}
			}

			if (_value is LuaObjectDescriptor)
			{
				return (_value as LuaObjectDescriptor).obj;
			}

			return _value;
		}

		/// <summary>
		/// 设置对象
		/// </summary>
		/// <param name="keyPath">键名路径.</param>
		/// <param name="value">值.</param>
		public void setObject(String keyPath, object value)
		{
			if (_context != null && _tableId != null && type == LuaValueType.Map)
			{
				LuaValue objectValue = new LuaValue (value);

				IntPtr resultPtr = IntPtr.Zero;

				IntPtr valuePtr = IntPtr.Zero;
				LuaObjectEncoder encoder = new LuaObjectEncoder (_context);
				encoder.writeObject (this);
				byte[] valueBytes = encoder.bytes;
				valuePtr = Marshal.AllocHGlobal (valueBytes.Length);
				Marshal.Copy (valueBytes, 0, valuePtr, valueBytes.Length);


				IntPtr objPtr = IntPtr.Zero;
				LuaObjectEncoder objEncoder = new LuaObjectEncoder (_context);
				objEncoder.writeObject (objectValue);

				byte[] bytes = objEncoder.bytes;
				objPtr = Marshal.AllocHGlobal (bytes.Length);
				Marshal.Copy (bytes, 0, objPtr, bytes.Length);

				int bufferLen = NativeUtils.tableSetObject (_context.objectId, valuePtr, keyPath, objPtr, resultPtr);

				if (bufferLen > 0)
				{
					LuaValue resultValue = LuaObjectDecoder.DecodeObject (resultPtr, bufferLen, _context) as LuaValue;
					_value = resultValue._value;
				}

				if (valuePtr != IntPtr.Zero)
				{
					Marshal.FreeHGlobal (valuePtr);
				}

				if (objPtr != IntPtr.Zero)
				{
					Marshal.FreeHGlobal (objPtr);
				}
			}
		}
	}
}